### ISSUE TITLE
sphinx.pycode.ast.unparse does not understand operator precedence

### ISSUE THREAD

**Issue by eric-wieser** [sphinx-doc/sphinx#7498]
As an example:
```python
>>> import sphinx.pycode.ast as ast
>>> ast.unparse(ast.parse("(1 + 2) * 3").body[0].value)
'1 + 2 * 3'
>>> ast.unparse(ast.parse("(1, 2) + (3, 4)").body[0].value)
'1, 2 + 3, 4'
```

Easiest way to fix this will be to copy some of the approach from the python 3.9 `ast.unparse`, which needs #7497 in order to keep track of the current precedence level.

**Comment by tk0miya** [comment]
Please mind we also need to support `Callable[[str, str], str]`. This type hint also uses a tuple without parenthesis.

**Comment by eric-wieser** [comment]
Indeed, the idea would be to add parentheses only where needed, like `ast.unparse` does

**Comment by eric-wieser** [comment]
Although note that `Callable[([str, str], str)]` is perfectly correct, it just is unfortunately also ugly.

**Comment by tk0miya** [comment]
It seems `ast.unparse()` uses `items_view()` to generate common part of sequences. It's cool.

**Comment by eric-wieser** [comment]
Copying the entirety of `ast.unparse` into sphinx may also be a sensible option.

**Comment by tk0miya** [comment]
Reasonable. But we should notice about license issue to copy it.

**Comment by tk0miya** [comment]
Fixed by #8265 now.
